main/src/ca/uwaterloo/flix/language/phase/njvm/Mnemonics.scala

```diff
@@ -359,940 +638,1496 @@ object Mnemonics {
       */
     def RETURN_VOID: F[StackNil] => F[StackNil] = t => t.emitReturnVoid()
 
+    /**
+      * Returns without value
+      */
+    def RETURN_VOID[T <: MnemonicsTypes : TypeTag](implicit root: Root, flix: Flix): F[StackNil ** T] => F[StackNil] = {
+      t => t.emitReturnVoid()
+    }
+
+
     /**
       * Pushes the result of adding the two top-most ints.
       */
-    def IADD[R <: Stack]: F[R ** PrimInt ** PrimInt] => F[R ** PrimInt] = ???
+    def IADD[S <: Stack]: F[S ** MInt ** MInt] => F[S ** MInt] = ???
 
     /**
       * Creates a new instance of the type parameter T which should be a subtype of JvmType.Reference and puts
       * it on top of the stack
       */
-    //TODO: T should be a subtype of JvmType.Reference not MnemonicsType
-    def NEW[R <: Stack](jt: Reference)(implicit root: Root, flix: Flix): F[R] => F[R ** Reference] =
+    def NEW[S <: Stack](jt: Reference)(implicit root: Root, flix: Flix): F[S] => F[S ** Ref[MObject]] =
       t => t.emitNew(jt)
 
     /**
       * Duplicates whatever is on Top of the stack
       */
-    def DUP[R <: Stack, T: TypeTag]: F[R ** T] => F[R ** T ** T] =
+    def DUP[S <: Stack, T: TypeTag]: F[S ** T] => F[S ** T ** T] =
       t => t.emitDup()
 
     /**
       * Loads a constant string onto the stack
       */
-    def LDC_STRING[R <: Stack](value: String): F[R] => F[R ** JString.type] =
+    def LDC_STRING[S <: Stack](value: String): F[S] => F[S ** Ref[MString]] =
       t => t.emitLdc(value)
 
+    /**
+      * Loads a constant int onto the stack
+      */
+    def LDC_INT[S <: Stack](value: Int): F[S] => F[S ** MInt] =
+      t => t.emitLdc(value.asInstanceOf[Object])
+
+    /**
+      * Loads a constant int onto the stack
+      */
+    def CONST_NULL[S <: Stack, T <: Ref[MObject]]: F[S] => F[S ** T] =
+      t => t.emitConstNull
+
+    /**
+      * Loads a constant int onto the stack
+      */
+    def NEWARRAY[S <: Stack, T <: MObject : TypeTag](implicit root: Root, flix: Flix): F[S ** MInt] => F[S ** MArray[Ref[T]]] =
+      t => t.emitNewArray(getJvmType[Ref[T]].asInstanceOf[Reference])
+
     /**
       * Throws the exception which is currently on top of the stack
       */
-    def THROW: F[StackNil ** Reference] => F[StackNil] =
+    def THROW: F[StackNil ** Ref[MObject]] => F[StackNil] =
       t => t.emitThrow()
 
-    def IFEQ[S <: Stack](f: F[S] => F[S]): F[S ** PrimBool] => F[S] = {
+    def IFEQ[S <: Stack](f: F[S] => F[S]): F[S ** MBool] => F[S] = {
       val skipLabel = new Label
-      ((t: F[S ** PrimBool]) => t.emitIfeq[S](skipLabel)) |>>
+      ((t: F[S ** MBool]) => t.emitIfeq[S](skipLabel)) |>>
         f |>>
         (t => t.emitLabel(skipLabel))
     }
 
-    def CHECK_CAST[S <: Stack](jt: Reference): F[S ** Reference] => F[S ** Reference] =
+    def CHECK_CAST[S <: Stack](jt: Reference): F[S ** Ref[MObject]] => F[S ** Ref[MObject]] =
```

Comment:
I would expect CHECK_CAST to actually change the type on the stack (i.e. to make it more precise)